pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -42,6 +57,360 @@ type Adapter struct {
+			klog.V(3).Infof("Manifest type: %s", gvr.String())
+
+			labels := map[string]string{
+				"component": applicationName,
+			}
+
+			manifestLabels := deploymentManifest.GetLabels()
+			if manifestLabels != nil {
+				for key, value := range labels {
+					manifestLabels[key] = value
+				}
+				deploymentManifest.SetLabels(manifestLabels)
+			} else {
+				deploymentManifest.SetLabels(labels)
+			}
+
+			// Check to see whether deployed resource already exists. If not, create else update
+			instanceFound := false
+			item, err := a.Client.DynamicClient.Resource(gvr).Namespace(namespace).Get(deploymentManifest.GetName(), metav1.GetOptions{})
+			if item != nil && err == nil {
+				instanceFound = true
+				deploymentManifest.SetResourceVersion(item.GetResourceVersion())
+				deploymentManifest.SetAnnotations(item.GetAnnotations())
+				if item.GetKind() == "Service" {
+					currentServiceSpec := item.UnstructuredContent()["spec"].(map[string]interface{})
+					if currentServiceSpec["type"] == "ClusterIP" {
+						newService := deploymentManifest.UnstructuredContent()
+						newService["spec"].(map[string]interface{})["clusterIP"] = currentServiceSpec["clusterIP"]
+						deploymentManifest.SetUnstructuredContent(newService)
+					}
+				}
+			}
+
+			s := log.Spinnerf("Deploying the manifest for %s", gvk.Kind)
+			result := &unstructured.Unstructured{}
+			actionType := "create"
+			if !instanceFound {
+				result, err = a.Client.DynamicClient.Resource(gvr).Namespace(namespace).Create(deploymentManifest, metav1.CreateOptions{})
+			} else {
+				actionType = "update" // Update deployment
+				result, err = a.Client.DynamicClient.Resource(gvr).Namespace(namespace).Update(deploymentManifest, metav1.UpdateOptions{})
+			}
+			if err != nil {
+				s.End(false)
+				return errors.Wrapf(err, "Failed to %s manifest %s", actionType, gvk.Kind)
+			} else {
+				s.End(true)
+				log.Successf("%sd manifest for %s (%s)", strings.Title(actionType), applicationName, gvk.Kind)
+			}
+
+			// Write the returned manifest to the local manifest file
+			if writtenToManifest {
+				_, err = manifestFile.WriteString("---\n")
+				if err != nil {
+					return errors.Wrap(err, "Unable to write to local manifest file")
+				}
+			}
+			err = yamlDecoder.Encode(result, manifestFile)
+			if err != nil {
+				return errors.Wrap(err, "Unable to write to local manifest file")
+			}
+			writtenToManifest = true
+		}
+	}
+
+	s := log.Spinner("Determining the application URL")
+
+	// TODO: Can we use a occlient created somewhere else rather than create another
+	client, err := occlient.New()
+	if err != nil {
+		return err
+	}
+
+	// Need to wait for a second to give the server time to create the artifacts
+	// TODO: Replace wait with a wait for object to be created
+	time.Sleep(2 * time.Second)
+
+	fullURL := ""
+	urlList, err := url.List(client, &config.LocalConfigInfo{}, "", applicationName)
```

Comment:
@groeges would the applicationName be the componentName here?

Reply:
No, applicationName := a.ComponentName + "-deploy"
The manifests I have been using have template values for "COMPONENT_NAME" and these are substituted with applicationName, so when we are searching we need to use applicationName.

Reply:
okay, but I was thinking more around the function definition: 
```[golang]
func List(client *occlient.Client, localConfig *config.LocalConfigInfo, componentName string, applicationName string) (URLList, error) {
```
Does applicationName in our code represent the componentName or the applicationName? The selectors used within the function are different because the labels are different

Reply:
It represents the applicationName. The url.List uses the applicationName to search for the route using app.kubernetes.io/part-of=\<applicationName\> and I had set these up in my manifests. 
We can change this if needed but would then not be able to use the url.List to find the route.

Reply:
So this means that the manifest from the stacks needs those labels and corresponding placeholders in order for odo to extract the url correctly?

Could we add the labels as part of the deploy?